### ISSUE TITLE
train_new_from_iterator consumes large amount of ram

### ISSUE THREAD

**Issue by RichardErkhov** [huggingface/tokenizers#1345]
When trying to train tokenizer on rather small dataset, training consumes quite a large amount of ram.
My dataset size is 2gb, but it takes a whole 150gb of ram.
Dataset link: [TinyStoriesV2-GPT4-train.txt](https://huggingface.co/datasets/roneneldan/TinyStories/blob/main/TinyStoriesV2-GPT4-train.txt)
code for reproduction:
```
from transformers import AutoTokenizer
file_path="TinyStoriesV2-GPT4-train.txt"
old_tokenizer = AutoTokenizer.from_pretrained("openlm-research/open_llama_3b")
def batched_data_generator(file_path, batch_size=1000):
    with open(file_path, 'r', encoding='utf-8') as file:
        batch = []
        for line in file:
            batch.append(line.strip())  # Remove newline characters
            if len(batch) == batch_size:
                yield batch
                batch = []
        if batch:  # Handle the last batch which might be smaller than batch_size
            yield batch
data_gen = batched_data_generator(file_path, 1024)
tokenizer = old_tokenizer.train_new_from_iterator(data_gen, 4096) #ram usage goes from few hundreds of megabytes to 150gb
```

os: ubuntu 22.04
python: 3.10.12

**Comment by RichardErkhov** [comment]
also should it take that long?
```
[00:02:06] Pre-processing sequences                 ████████ 0        /        0
[00:05:41] Tokenize words                           ████████ 11708602 / 11708602
[00:08:50] Count pairs                              ████████ 11708602 / 11708602
[00:34:51] Compute merges                           ░░░░░░░░ 80       /     4096
```

**Comment by ArthurZucker** [comment]
Hey! I think you should leverage the `Dataset` classes instead of writing a custom for loop like this, which is most probably the source of the error. 
Something like this should be better:
```python
import torch
from transformers import AutoTokenizer, BatchEncoding

file_path = "TinyStoriesV2-GPT4-train.txt"
old_tokenizer = AutoTokenizer.from_pretrained("openlm-research/open_llama_3b")

class TextDataset(torch.utils.data.Dataset):
    def __init__(self, file_path, batch_size):
        self.batch_size = batch_size
        self.lines = []
        
        with open(file_path, 'r', encoding='utf-8') as file:
            for line in file:
                self.lines.append(line.strip())  # Remove newline characters

    def __len__(self):
        return len(self.lines)

    def __getitem__(self, idx):
        batch = self.lines[idx:idx + self.batch_size]
        return batch

# Create the dataset, and process the full file. 
dataset = TextDataset(file_path, batch_size=1024)

# DataLoader for efficient batch processing
dataloader = torch.utils.data.DataLoader(dataset, batch_size=None)

# Train the new tokenizer
tokenizer = old_tokenizer.train_new_from_iterator(dataloader, 4096)
```
The file processing should not happen on the fly.

**Comment by Narsil** [comment]
Also super long lines can really affect overall memory usage. Could you try filtering out long lines.

**Comment by RichardErkhov** [comment]
> Hey! I think you should leverage the `Dataset` classes instead of writing a custom for loop like this, which is most probably the source of the error. Something like this should be better:
> 
> ```python
> import torch
> from transformers import AutoTokenizer, BatchEncoding
> 
> file_path = "TinyStoriesV2-GPT4-train.txt"
> old_tokenizer = AutoTokenizer.from_pretrained("openlm-research/open_llama_3b")
> 
> class TextDataset(torch.utils.data.Dataset):
>     def __init__(self, file_path, batch_size):
>         self.batch_size = batch_size
>         self.lines = []
>         
>         with open(file_path, 'r', encoding='utf-8') as file:
>             for line in file:
>                 self.lines.append(line.strip())  # Remove newline characters
> 
>     def __len__(self):
>         return len(self.lines)
> 
>     def __getitem__(self, idx):
>         batch = self.lines[idx:idx + self.batch_size]
>         return batch
> 
> # Create the dataset, and process the full file. 
> dataset = TextDataset(file_path, batch_size=1024)
> 
> # DataLoader for efficient batch processing
> dataloader = torch.utils.data.DataLoader(dataset, batch_size=None)
> 
> # Train the new tokenizer
> tokenizer = old_tokenizer.train_new_from_iterator(dataloader, 4096)
> ```
> 
> The file processing should not happen on the fly.

![image](https://github.com/huggingface/tokenizers/assets/104026872/092440bd-b0df-4304-8bd8-ca43d972c811)
Im not sure what's happening, I think it's looping over and over again, Im not sure why this is happening, it's consuming 98gb of ram already, I don't think it suppose to be like that.
Also I don't think it's the best thing to load the full dataset into the memory, because there are some datasets that are 1TB big, and it's for sure not going to fit into the memory

**Comment by ArthurZucker** [comment]
I don't know what you are doing, but the snippet I shared takes seconds to run for me

**Comment by morphpiece** [comment]
So for someone who comes for same issue of memrory hogging, the thing that worked for me was to change the old tokenizer, from `meta-llama/Llama-2-7b-hf` to `meta-llama/Meta-Llama-3-8B`. Not exactly sure why it would make a difference, but it did; by orders of magnitude!

**Comment by ArthurZucker** [comment]
It's not the same tokenizer ! (Llama2 most notably uses a normalizer and no pre_tokenizer, while llama3 does not, but both should use the pre_tokenizer. It's contrallable with the `legacy=False` flag in `transformers`)

**Comment by kdcyberdude** [comment]
Hi @ArthurZucker, I am also facing the same issue, Can you help me understand the reason behind this and provide any references or suggestions to train this model more efficiently?
```python
from datasets import load_from_disk, load_dataset
from transformers import AutoTokenizer

ds = load_dataset('kdcyberdude/Vichaar', num_proc=8, cache_dir='./gemma_data_cache')['train']
print(ds)
tokenizer = AutoTokenizer.from_pretrained("openchat/openchat-3.5-0106-gemma")

def batch_iterator(batch_size=1000):
    for i in range(0, len(ds), batch_size):
        yield ds[i : i + batch_size]["text"]

new_tokenizer = tokenizer.train_new_from_iterator( batch_iterator(), vocab_size=32_000, length=len(ds))
new_tokenizer.save_pretrained("./gemma-32k-pa-tokenizer")
```
![Screenshot from 2024-06-30 03-02-24](https://github.com/huggingface/tokenizers/assets/34835322/0e09168e-5d43-4868-b15e-bafdd177e591)

I am getting OOM even in 1TB RAM instance. I have also tried this using data loader, the `Pre-processing sequences` steps keeps on iterating even after len(ds) and memory keeps increasing. The iteration goes 7*len(ds) until it hits OOM. Not sure when it will stop. Same as https://github.com/huggingface/tokenizers/issues/1434 
```python
class TextDataset(torch.utils.data.Dataset):
    def __init__(self, ds, batch_size):
        self.batch_size = batch_size
        self.ds = ds

    def __len__(self):
        return len(self.ds)

    def __getitem__(self, idx):
        batch = self.ds[idx:idx + self.batch_size]['text']
        return batch

dataset = TextDataset(ds, batch_size=1024)
dataloader = torch.utils.data.DataLoader(dataset, batch_size=None)

new_tokenizer = tokenizer.train_new_from_iterator( dataloader, vocab_size=32_000, length=len(ds))
```
I also tried debugging the code to understand which part is consuming this much RAM but I am not able to get into this `train_from_iterator` function in [`tokenization_utils_fast.py`](https://github.com/huggingface/transformers/blob/e65502951593a76844e872fee9c56b805598538a/src/transformers/tokenization_utils_fast.py#L817). I am speculating this could be a calling an executable/binary code may be running in rust. 

Any help or pointers would be greatly appreciated!
